libsolidity/ast/Types.cpp

```diff
@@ -2135,55 +2134,93 @@ MemberList::MemberMap StructType::nativeMembers(ContractDefinition const*) const
 	return members;
 }
 
-TypePointer StructType::interfaceType(bool _inLibrary) const
+TypeResult StructType::interfaceType(bool _inLibrary) const
 {
 	if (_inLibrary && m_interfaceType_library.is_initialized())
 		return *m_interfaceType_library;
 
 	if (!_inLibrary && m_interfaceType.is_initialized())
 		return *m_interfaceType;
 
-	TypePointer result = TypePointer{};
+	TypeResult result{TypePointer{}};
 
-	if (_inLibrary && location() == DataLocation::Storage)
-		result = shared_from_this();
-	else if (recursive())
-		result = TypePointer{};
-	else
+	m_recursive = false;
+
+	auto visitor = [&](
+		StructDefinition const& _struct,
+		CycleDetector<StructDefinition>& _cycleDetector,
+		size_t /*_depth*/
+	)
 	{
 		// Check that all members have interface types.
-		// We pass "false" to canBeUsedExternally (_inLibrary), because this struct will be
-		// passed by value and thus the encoding does not differ, but it will disallow
-		// mappings.
-		// Also return false if at least one struct member does not have a type.
+		// Return false if at least one struct member does not have a type.
 		// This might happen, for example, if the type of the member does not exist,
 		// which is reported as an error.
-		bool allOkay = true;
-		for (auto const& var: m_struct.members())
+		for (ASTPointer<VariableDeclaration> const& variable: _struct.members())
 		{
 			// If the struct member does not have a type return false.
 			// A TypeError is expected in this case.
-			if (!var->annotation().type)
+			if (!variable->annotation().type)
 			{
-				allOkay = false;
-				break;
+				result = TypeResult::err("Invalid type!");
+				return;
 			}
-			if (!var->annotation().type->interfaceType(false))
+
+			Type const* memberType = variable->annotation().type.get();
+
+			while (dynamic_cast<ArrayType const*>(memberType))
+				memberType = dynamic_cast<ArrayType const*>(memberType)->baseType().get();
+
+			if (StructType const* innerStruct = dynamic_cast<StructType const*>(memberType))
+				if (_cycleDetector.run(innerStruct->structDefinition()))
```

Comment:
```suggestion
				if (innerStruct->m_recursive == true && _cycleDetector.run(innerStruct->structDefinition()))
```

Reply:
ah ,that's what you meant. Got it. :+1:

Reply:
Can you add this? Otherwise the PR is fine.

Reply:
I thought I did but it turned out I didn't save the file :1st_place_medal: